deepchem/models/torch_models/grover.py

```diff
@@ -92,3 +98,312 @@ def forward(self, graph_batch: List[torch.Tensor]):
+        self.model = self.build_model(hidden_size, dropout, activation,
+                                      num_mt_block, num_attn_head, bias)
+        self.components = self.build_components(hidden_size, atom_vocab_size,
+                                                bond_vocab_size, fg_size)
+
+    def build_model(self, hidden_size, dropout, activation, num_mt_block,
+                    num_attn_head, bias):
+        return GroverEmbedding(embedding_output_type='both',
+                               hidden_size=hidden_size,
+                               dropout=dropout,
+                               activation=activation,
+                               num_mt_block=num_mt_block,
+                               num_attn_head=num_attn_head,
+                               bias=bias)
+
+    def build_components(self, hidden_size, atom_vocab_size, bond_vocab_size,
+                         fg_size):
+        components = {}
+        # layer for predicting atom vocabulary using atom emeddings
+        components['atom_vocab_task_atom'] = self._atom_vocab_prediction_model(
+            atom_vocab_size, hidden_size)
+        # layer for predicting atom vocabulary using bond emeddings
+        components['atom_vocab_task_bond'] = self._atom_vocab_prediction_model(
+            atom_vocab_size, hidden_size)
+        # layer for predicting bond vocabulary using atom embeddings
+        components['bond_vocab_task_atom'] = _BondVocabPredictionLayer(
+            bond_vocab_size, hidden_size)
+        # layer for predicting bond vocabulary using bond embeddings
+        components['bond_vocab_task_bond'] = _BondVocabPredictionLayer(
+            bond_vocab_size, hidden_size)
+        # layer for predicting functional group
+        components['fg_task_all'] = _FunctionalGroupPrediction(fg_size)
+        return components
+
+    def _atom_vocab_prediction_layer(self, vocab_size, hidden_size):
+        return nn.Sequential(nn.Linear(hidden_size, vocab_size),
+                             nn.LogSoftmax(dim=1))
+
+    def loss_func(self, inputs, labels, weights):
+        """Loss function for grover pretrainer"""
+        dist_coff = 0.1
+        embeddings = self.model(inputs)
+
+        av_task_atom_pred = self.components['atom_vocab_task_atom'](
+            embeddings["atom_from_atom"])
+        av_task_bond_pred = self.components['atom_vocab_task_bond'](
+            embeddings["atom_from_bond"])
+
+        bv_task_atom_pred = self.components['bond_vocab_task_atom'](
+            embeddings["bond_from_atom"])
+        bv_task_bond_pred = self.components['bond_vocab_task_bond'](
+            embeddings["bond_from_bond"])
+
+        fg_task_pred_all = self.components['fg_task_all'](embeddings,
+                                                          atom_scope,
+                                                          bond_scope)
+
+        preds = {}
+        preds['av_task'] = (av_task_atom_pred, av_task_bond_pred)
```

Comment:
This function is really complicated; let's discuss offline how we can simplify this implementation